From 3562c6d19f341d803c33f8498fa0869ac7d9dea6 Mon Sep 17 00:00:00 2001
From: Lu Wei <luwei32@huawei.com>
Date: Thu, 13 Apr 2023 22:02:06 +0800
Subject: [PATCH 0044/1646] [Huawei] net/ns: Fix a soft lockup in cleanup_net

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 186807

-------------------------------

When namespaces is cleaned up, every netdevice should be notified. If
there is too many devices, it will take will take too many CPU time to
cause a soft lockup as the following trace:

watchdog: BUG: soft lockup - CPU#1 stuck for 23s!
...
Call Trace:
 addrconf_notify+0x5b/0x680
 raw_notifier_call_chain+0x44/0x60
 unregister_netdevice_many+0x1d0/0x460
 default_device_exit_batch+0x141/0x170
 cleanup_net+0x168/0x290
 process_one_work+0x1ab/0x340
 worker_thread+0x49/0x2f0
 kthread+0xfe/0x140
 ret_from_fork+0x22/0x30

The reproducer:
`
ns=dummy-ns
NIFS=50

for ((i=0; i<$NIFS; i++))
do
        ip netns add dummy-ns$i
        ip netns exec dummy-ns$i ip link set lo up
done

for ((j=0; j<$NIFS; j++))
do
        for ((i=0; i<5000; i++))
        do
                if=eth$j$i
                ip netns exec dummy-ns$j ip link add $if type dummy
                ip netns exec dummy-ns$j ip link set $if up
        done
done

for ((i=0; i<$NIFS; i++))
do
        ip netns del dummy-ns$i
done
`

This patch calls cond_resched() to give the scheduler a chance to run a
higher-priority process

Fixes: 8e4d080c9a8a ("net: move rollback_registered_many()")
Signed-off-by: Lu Wei <luwei32@huawei.com>
---
 net/core/dev.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/net/core/dev.c b/net/core/dev.c
index 33244d3f940d..c1aa94b013dc 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -10769,6 +10769,7 @@ void unregister_netdevice_many(struct list_head *head)
 		/* Remove XPS queueing entries */
 		netif_reset_xps_queues_gt(dev, 0);
 #endif
+		cond_resched();
 	}
 
 	synchronize_net();
-- 
2.17.1

